File: rd/trunk/Client/cms_lua/Assets/StreamingAssets/lua/logic/Battle/BattleController.lua
require "logic.Battle.BattleProcess"

BattleController = {
	battleGroup = nil,
	curProcess = nil,
}

testProcess = 1
processCount = 0

function BattleController:StartBattle(group)
	self.battleGroup = group

	-- create all units of this battle
	self:CreateUnit()

	-- 前置剧情动画
	self:PlayPreStoryAnim()

	-- 显示对局UI
	self:ShowUI()

	-- 开始进程
	self:StartProcess()

end

function BattleController:CreateUnit()
	print("<Battle>Creating Battle Units...")
end

function BattleController:PlayPreStoryAnim()
	print("<Battle>Play Story Animation before any process...")
end

function BattleController:PlayPostStoryAnim()
	print("<Battle>Play Story Animation after any process...")
end

function BattleController:StartProcess()
	self.curProcess = self:GetNextProcess()
	if self.curProcess ~= nil then
		self.curProcess:Start(self.battleGroup, function ()
			self:OnEndProcess()
		end)
	else
		self:OnEndAllProcess();
	end
end

function BattleController:OnEndProcess()
	print("<Battle>OnEndProcess Callback")
	self:StartProcess();
end

function BattleController:OnEndAllProcess()
	print("<Battle>OnEndAllProcess")
	-- 胜利失败动画
	self:PlayBalanceAnim()

	-- 后置剧情动画
	self:PlayPostStoryAnim()

	-- 结算面板
	self:ShowBalanceUI()
end

function BattleController:ShowUI( ... )
	print("<Battle>Show Battle UI...")
end

function BattleController:GetNextProcess( ... )
	if processCount < testProcess then	
		processCount = processCount+1
		return BattleProcess:New()
	else
		return nil
	end
end

function BattleController:PlayBalanceAnim( ... )
	print("<Battle>Playing Balance Animation...")
end

function BattleController:ShowBalanceUI( ... )
	print("<Battle>Show the balance UI of this battle...")
end

-----------------------------------------Event-----------------------------------------------------
function BattleController:OnSkipAnim(  )
	-- body
end

File: rd/branch/Client-qa/cms/Assets/script/Battle/BattleController.cs
using System;
using System.Collections.Generic;
using UnityEngine;
using UnityEngine.EventSystems;

public enum BattleType
{
    Normal,
    Boss,
    Rare,
}

public class BattleController : MonoBehaviour
{
    BattleType battleType;
    InstanceData instanceData;
    BattleProcess process;
    bool isMouseOnUI = false;

    BattleGroup battleGroup;
    public BattleGroup BattleGroup
    {
        get { return battleGroup; }
    }

    //初值为-1
    int curProcessIndex = -1;

    static BattleController instance;
    public static BattleController Instance
    {
        get { return instance; }
    }


    // Use this for initialization
    public void Init(BattleProcess process)
    {
        instance = this;
        this.process = process;
        battleGroup = new BattleGroup();
    }

    void Update()
    {
        RaycastHit hit;
        Ray ray = BattleCamera.Instance.GetComponent<Camera>().ScreenPointToRay(Input.mousePosition);
        if (Input.GetMouseButtonDown(0))
        {
            isMouseOnUI = EventSystem.current.IsPointerOverGameObject();
        }

        if (Input.GetMouseButtonUp(0) && !isMouseOnUI)
        {
            if (Physics.Raycast(ray, out hit, 100))
            {
                var battleGo = hit.collider.gameObject.GetComponent<BattleObject>();
                if (battleGo)
                {
                    OnHitBattleObject(battleGo);
                }
                else
                {
                    Logger.LogWarning("Hit something but not a battle object!");
                }
            }
            else
            {
                GameEventMgr.Instance.FireEvent(GameEventList.HideSwitchPetUI);
            }
        }
    }

    void OnHitBattleObject(BattleObject battleGo)
    {
        if (battleGo.camp == UnitCamp.Enemy)
        {
            //设置集火目标
            process.OnHitBattleObject(battleGo);
        }
        else if (battleGo.camp == UnitCamp.Player)
        {
            if (process.CanSwitchPet())
            {
                //换宠
                ShowSwitchPetUIArgs args = new ShowSwitchPetUIArgs();
                args.targetId = battleGo.id;
                args.idleUnits = battleGroup.GetPlayerOffsiteUnits();
                GameEventMgr.Instance.FireEvent<EventArgs>(GameEventList.ShowSwitchPetUI, args);
            }
        }
    }

    public void StartBattle(PbStartBattle proto)
    {
        battleType = (BattleType)proto.battleType;
        instanceData = StaticDataMgr.Instance.GetInstanceData(proto.instanceId);

        //设置battlegroup 并且创建模型
        battleGroup.SetEnemyList(proto.enemyList);
        battleGroup.SetPlayerList(proto.playerList);

        PlayPreStoryAnim();

        ShowUI();

        StartProcess();
    }

    void PlayPreStoryAnim()
    {
        Logger.Log("[Battle]Play Story Animation before any process...");
    }

    void ShowUI()
    {
        GameEventMgr.Instance.FireEvent(GameEventList.ShowBattleUI);
    }

    void StartProcess()
    {
        var curProcess = GetNextProcess();
        if (curProcess != null)
            process.StartProcess(curProcess);
        else
            OnAllProcessOver();
    }

    ProcessData GetNextProcess()
    {
        curProcessIndex++;

        if (battleType == BattleType.Normal)
        {
            if (curProcessIndex == 0)
                return new ProcessData();
            else
                return null;
        }

        if (battleType == BattleType.Boss)
        {
            if (curProcessIndex >= instanceData.bossProcess.Count)
                return null;
            else
                return null;
        }

        if (battleType == BattleType.Rare)
        {
            if (curProcessIndex >= instanceData.rareProcess.Count)
                return null;
            else
                return null;
        }

        Logger.LogError("BattleType error" + battleType);
        return null;
    }

    public void OnProcessSuccess()
    {
        StartProcess();
    }

    public void OnProcessFailed()
    {
        OnAllProcessOver();
    }

    public void OnAllProcessOver()
    {
        //胜利失败动画
        PlayBalanceAnim();
        //后置剧情动画
        PlayPostStoryAnim();
        //结算面板UI
        ShowBalanceUI();
    }

    private void PlayBalanceAnim()
    {
        Logger.Log("[Battle]Playing Balance Anim...");
    }

    private void PlayPostStoryAnim()
    {
        Logger.Log("[Battle]Playing Post Story Anim...");
    }

    private void ShowBalanceUI()
    {
        Logger.Log("[Battle]Showing Balance UI...");
    }
}
